# Perceptron Neural Network Implementation

Renato Barros

## Introduction

In the dynamic landscape of the corporate world, the ability to make intelligent
and strategic decisions is essential for success. In this context, the use of
the Perceptron—a machine learning algorithm inspired by the functioning of
biological brain neurons—can serve as a powerful tool for data analysis and
behavior prediction.

The Perceptron is like a dedicated learner that, by observing patterns and
relationships in data, learns to make predictions and assist in decision-making.
Whether identifying products with high sales potential, detecting fraudulent
activities, or personalizing customer experiences, it offers a broad range of
applications to drive efficiency and intelligence in modern organizations.

## How It Works

The Perceptron operates as a foundational artificial neural network: it receives
input signals, processes them, and produces an output. This processing involves
straightforward mathematical operations:

1.  **Linear Combination:** Each input signal (represented by a numerical value)
    is multiplied by a corresponding weight, which denotes its importance in
    the final decision. The products are summed together, yielding an
    intermediate value known as the "weighted sum".

    Formula: `u = (x1 * weight1) + (x2 * weight2) + ...`

2.  **Activation Function:** The weighted sum (u) computed in the previous step
    is passed through an activation function, which determines the final output
    of the Perceptron. The activation function introduces non-linearity to the
    model, mapping inputs to categorical decisions. In our case, we use the
    "step function" (Heaviside step function), which works like a switch:

    Formula:

    `f(u) = 1 if u >= 0 else 0`

    Where:

    -   `f(u)` is the activation function output.
    -   `u` is the weighted sum computed in the previous step.

    In other words, if the weighted sum is greater than or equal to zero, the
    output is 1; otherwise, the output is 0. This binary output can represent,
    for instance, recommending or not recommending a product, classifying a
    user as a prospective buyer, or any target binary event.

3.  **Delta Rule:** After calculating the output using the activation function,
    the Perceptron compares this prediction with the target output (the ground
    truth reference value). The difference between the desired target and the
    actual output is the "error". To minimize this error and learn from mistakes,
    the Perceptron adjusts its weights using the delta rule (Widrow-Hoff rule).
    This rule updates each weight proportionally to the error and the corresponding input.

    Formula:
    `new_weight = current_weight + (learning_rate * error * input)`

    Where:

    -   `learning_rate` is a small positive factor controlling learning speed
        (as defined in the "Model Parameters & Training Data" section).

To illustrate the Perceptron in practice, the following Python implementation
builds a clean model using corporate case examples and analogies.

## Model Parameters & Training Data

Just like an apprentice requires initial guidance, the Perceptron needs initial
hyperparameters and training examples to begin its learning journey.

In [1]:
# Perceptron Neural Network Implementation
# Copyright (C) 2026 Renato Barros
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.

# Initialize perceptron weights (w1 and w2)
peso = [0.2, -0.1]
# Define the learning rate for the algorithm
taxa_aprendizagem = 0.1
# Initialize total error accumulator as zero
erros_totais = 0.0
# Define the maximum number of epochs (iterations) for training
limite_epocas = 20

# Define training dataset with input features (x1, x2) and desired target (ground truth)
dados_treinamento = [
    {"x1": 0.5, "x2": 0.8, "saida_desejada": 1},
    {"x1": 0.2, "x2": 0.4, "saida_desejada": 0},
]

-   **Initializing Weights:** In a Perceptron, "weights" represent the relative
    importance of each feature in the decision-making process. In the code,
    `peso = [0.2, -0.1]` initializes random starting weights for two relevant
    input signals (x1 and x2).

    For example, suppose we are building a model to predict whether a product
    will become a sales hit. Here, x1 could represent the number of positive
    reviews and x2 the number of times it was added to customer wishlists.
    The weights (0.2 and -0.1) reflect the relative influence of each feature.

-   **Setting the Learning Rate:** `taxa_aprendizagem = 0.1` controls how
    aggressively the Perceptron updates its weights upon encountering an error.
    Think of the learning rate like the sensitivity of a scale.

    A high learning rate makes large weight adjustments, accelerating training
    but risking instability or oscillation. A low learning rate makes smaller,
    gradual adjustments, resulting in slower but more stable convergence.

-   **Accumulating Errors:** `erros_totais = 0.0` acts as a global error counter.
    During training, the Perceptron compares predictions with ground truth,
    accumulating errors to track how well the model is learning.

    Imagine training a Perceptron to identify prospective buyers on an e-commerce
    platform. Each time a user is misclassified, the error count increments,
    providing a direct metric of accuracy.

-   **Limiting Training Epochs:** `limite_epocas = 20` defines the maximum
    number of full passes over the entire training dataset, preventing infinite loops.

    In a model estimating delivery timeliness based on logistics features,
    each epoch represents analyzing all historical delivery records,
    allowing the weights to adjust iteratively toward higher precision.

-   **Providing Training Data:** `dados_treinamento` forms the foundation of
    learning: a collection of feature samples alongside their correct target
    labels (`saida_desejada = 1` or `0`).

    If training to predict viral products, the dataset contains past viral items
    contrasted against items that did not achieve viral status.

## Training the Perceptron

With parameters set and training data prepared, the Perceptron begins its learning
process. This iterative fine-tuning process searches for optimal weight values to
maximize prediction accuracy.

In [2]:
# Start Perceptron training loop
for epoca in range(1, limite_epocas + 1):
    print(f"\n Época: {epoca}\n")
    # Flag indicating whether an error occurred in current epoch (initially False)
    erro_na_epoca = False

    # Iterate over each training sample
    for dado in dados_treinamento:
        # Extract input features and target ground truth label
        x1 = dado["x1"]
        x2 = dado["x2"]
        saida_desejada = dado["saida_desejada"]

        # Read current weight values
        w1 = peso[0]
        w2 = peso[1]

        # Compute perceptron linear combination (u)
        u = (x1 * peso[0]) + (x2 * peso[1])
        # Apply step activation function:
        # if u >= 0, output = 1; otherwise, output = 0
        valor_saida = 1 if u >= 0 else 0

        # Compute error: target value minus perceptron output
        erro = saida_desejada - valor_saida

        # Accumulate absolute error
        erros_totais += abs(erro)

        # If an error occurred in this iteration, set flag to True
        if erro != 0:
            erro_na_epoca = True
            # Update weights (w1 and w2) using the delta rule
            peso[0] += taxa_aprendizagem * erro * x1
            peso[1] += taxa_aprendizagem * erro * x2

        # Display current iteration details (epoch, inputs, output, weights)
        print(
            f" X1: {x1:.1f}, X2: {x2:.1f} | W1: {w1:.2f}, W2: {w2:.2f} | Y: {
                saida_desejada}, "
            f"Y-predito: {valor_saida} | Erro: {
                erro}, Erro acumulado: {erros_totais}"
        )
        if erro_na_epoca:
            print(
                f" Ajuste de pesos: Novo W1 = {
                    peso[0]:.2f}, Novo W2 = {peso[1]:.2f}\n"
            )
        if not erro_na_epoca:
            print(" Ajuste de pesos: Não foi necessário.\n")
    print("-=-" * 31)

    # Stopping condition: if no error occurred throughout the epoch (erro_na_epoca = False)
    if not erro_na_epoca:
        # Print success message showing the epoch in which zero error was reached
        print(
            f"\n Modelo treinado com sucesso na época {
                epoca}! Erro zero atingido."
        )
        # Print final optimal weights (w1 and w2)
        print(
            f" Valores ideais de peso: W1 = {
                peso[0]:.2f}, W2 = {peso[1]:.2f}\n"
        )
        # Terminate training loop
        break
# If loop exhausts all epochs without zero error
else:
    # Print notification that epoch limit was reached
    print(
        f"\n O limite de {
            limite_epocas} épocas foi atingido sem alcançar erro zero."
    )
    # Indicate that perceptron failed to converge
    print(" O perceptron não pôde ser treinado com sucesso.\n")


 Época: 1

 X1: 0.5, X2: 0.8 | W1: 0.20, W2: -0.10 | Y: 1, Y-predito: 1 | Erro: 0, Erro acumulado: 0.0
 Ajuste de pesos: Não foi necessário.

 X1: 0.2, X2: 0.4 | W1: 0.20, W2: -0.10 | Y: 0, Y-predito: 1 | Erro: -1, Erro acumulado: 1.0
 Ajuste de pesos: Novo W1 = 0.18, Novo W2 = -0.14

-=--=--=--=--=--=--=--=--=--=--=--=--=--=--=--=--=--=--=--=--=--=--=--=--=--=--=--=--=--=--=-

 Época: 2

 X1: 0.5, X2: 0.8 | W1: 0.18, W2: -0.14 | Y: 1, Y-predito: 0 | Erro: 1, Erro acumulado: 2.0
 Ajuste de pesos: Novo W1 = 0.23, Novo W2 = -0.06

 X1: 0.2, X2: 0.4 | W1: 0.23, W2: -0.06 | Y: 0, Y-predito: 1 | Erro: -1, Erro acumulado: 3.0
 Ajuste de pesos: Novo W1 = 0.21, Novo W2 = -0.10

-=--=--=--=--=--=--=--=--=--=--=--=--=--=--=--=--=--=--=--=--=--=--=--=--=--=--=--=--=--=--=-

 Época: 3

 X1: 0.5, X2: 0.8 | W1: 0.21, W2: -0.10 | Y: 1, Y-predito: 1 | Erro: 0, Erro acumulado: 3.0
 Ajuste de pesos: Não foi necessário.

 X1: 0.2, X2: 0.4 | W1: 0.21, W2: -0.10 | Y: 0, Y-predito: 1 | Erro: -1, Erro acumu

### Code Walkthrough:

The code above represents the core training cycle of the Perceptron, an iterative
process that refines model weights epoch by epoch to minimize prediction error.
Let's break down each step:

1.  **Epoch Control:** The outer `for` loop manages training epochs, where each
    epoch is a complete pass over the entire dataset to compute errors and adjust
    weights. The maximum epoch limit (20) bounds these cycles.

2.  **Iterating over Training Data:** The inner `for` loop processes each sample
    individually, allowing the model to examine features (x1, x2) and ground truth
    (`saida_desejada`).

3.  **Feature Extraction:** For each sample, the code extracts inputs (x1, x2) and
    target label (`saida_desejada`) to compute the output and required updates.

4.  **Storing Current Weights:** The variables `w1` and `w2` hold current weight
    values for logging and display during training.

5.  **Computing Perceptron Output:** Evaluates linear combination `u = x1*w1 + x2*w2`.
    In our product popularity scenario, this represents the model's aggregated score.

6.  **Applying Activation Function:** Applies the step function: if `u >= 0`, output
    is 1 (success); otherwise 0 (failure).

7.  **Computing Error:** Calculates `error = saida_desejada - valor_saida`.

8.  **Weight Updates (Delta Rule):** If an error occurs, weights are adjusted by
    `taxa_aprendizagem * erro * x` to steer the decision boundary closer to correct classification.

9.  **Logging Iteration Details:** Displays epoch number, inputs, current weights,
    predicted output, error, and applied adjustments for full visibility.

10. **Stopping Criterion:** At the end of each epoch, if no errors occurred (`not erro_na_epoca`),
    training terminates early, reporting convergence.

The Perceptron training process seeks an equilibrium where all training samples
are correctly classified. Although simple, this mechanism forms the conceptual
foundation of modern deep learning.

## Interpretation of Results

Analyzing the training execution output allows us to assess convergence and
evaluate model learning. The training succeeds when the model achieves zero error
across all training samples.

As observed in the output logs, the Perceptron achieved zero error at epoch 12.
This indicates 12 passes were required to find a separating hyperplane.

Requiring multiple epochs demonstrates that while the data is linearly separable,
the initial random weights (`[0.2, -0.1]`) were not immediately optimal,
requiring progressive iterative adjustments.

At each step, weights shifted in response to classification errors, drawing closer
to the optimal boundary until complete separation occurred at epoch 12.

The final weights (`W1 = 0.23` and `W2 = -0.14`) capture the acquired knowledge:

-   **Positive W1 (+0.23):** Feature x1 has a positive influence on the target outcome.
    Higher x1 values increase the likelihood of outputting class 1.

-   **Negative W2 (-0.14):** Feature x2 has a negative influence on the target outcome.
    Higher x2 values decrease the likelihood of outputting class 1, acting as a counterbalance.

This demonstrates the model's ability to balance conflicting signals and learn meaningful relations.

Notably, if max epochs had been set to 3, training would have stopped prematurely
with suboptimal weights, resulting in poor classification and impaired generalization.

## Conclusion

Even in its elementary form, the Perceptron demonstrates the power of machine learning
to extract meaningful patterns from data and solve classification problems. From customer
segmentation to demand prediction and security screening, its principles lay the
groundwork for building intelligent, adaptive business systems.